src/Dhall/TypeCheck.hs

```diff
@@ -3254,3 +3256,18 @@ instance (Buildable a, Buildable s) => Buildable (DetailedTypeError s a) where
         source = case expr of
             Note s _ -> build s
             _        -> mempty
+
+{-| This function verifies that a custom context is well-formed so that
+    type-checking will not loop
+-}
+checkContext :: Context (Expr s X) -> Either (TypeError s X) ()
```

Comment:
I think you meant to export this. If not, please export it :smile:

Reply:
Actually, I intentionally did not export it since `typeWith` calls it once before entering the type-checking loop.  This avoids the performance issue since it's not called on every iteration of the loop